File: app/services/ai_service.py
import google.generativeai as genai
import json
import re
import logging
import time
import uuid
import os
import firebase_admin
from firebase_admin import credentials, firestore
from typing import Any, Dict, Type, TypeVar, Optional
from pydantic import BaseModel, ValidationError
from fastapi import HTTPException, status
from PIL import Image
import io
import base64
from app.core.ai_config import get_model, MODEL_NAME
from app.schemas.job import ExtractedJob, GeneratedApplication, SourceType, ExtractionConfidence
from app.schemas.analysis import JobAnalysis, AnalyzeJobRequest, Verdict, ApplyRecommendation, Confidence
from app.schemas.job import GenerateApplicationRequest
from app.services.web_fetcher import fetch_job_page
from app.services.ocr_service import extract_text_from_image

# Set up logging
logger = logging.getLogger(__name__)

# Initialize Firebase Admin
try:
    if not firebase_admin._apps:
        # In this environment, we can usually initialize without explicit credentials
        # as it uses the service account of the environment.
        firebase_admin.initialize_app()
    db = firestore.client()
except Exception as e:
    logger.error(f"Failed to initialize Firebase Admin: {e}")
    db = None

T = TypeVar("T", bound=BaseModel)

def extract_json(text: str) -> dict:
    """
    Find JSON inside response.
    Handles ```json blocks, raw JSON, and surrounding text.
    """
    try:
        # Clean up potential markdown artifacts
        text = text.strip()
        json_match = re.search(r"```json\s*(.*?)\s*```", text, re.DOTALL)
        if json_match:
            return json.loads(json_match.group(1))
        
        start = text.find('{')
        end = text.rfind('}')
        if start != -1 and end != -1:
            json_str = text[start:end+1]
            return json.loads(json_str)
            
        return json.loads(text)
    except Exception as e:
        logger.error(f"Failed to extract JSON from AI response. Error: {e}. Raw text: {text[:200]}...")
        raise ValueError(f"Invalid AI response format. Expected JSON, got: {text[:50]}...")

class AIService:
    def __init__(self):
        self.model = get_model()
        if not self.model:
            logger.error("CRITICAL: GEMINI_API_KEY is missing. AI Fit Match will fail.")

    def _safe_parse(self, data: Dict[str, Any], model_class: Type[T], request_id: str) -> T:
        """Normalize data and parse into Pydantic model with 422 error handling."""
        try:
            return model_class(**data)
        except ValidationError as e:
            logger.error(f"[{request_id}] Schema validation failed for {model_class.__name__}: {e}")
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail={
                    "error": "AI Schema Error",
                    "detail": f"AI returned invalid schema for {model_class.__name__}",
                    "request_id": request_id,
                    "code": "AI_PARSE_ERROR",
                    "fields": [str(err["loc"]) for err in e.errors()]
                }
            )

    def _log_interaction(self, action: str, prompt: str, response: str, latency: float, request_id: str, user_id: str = "anonymous", tokens_input: int = 0, tokens_output: int = 0):
        """
        Log AI interaction for auditing and troubleshooting.
        """
        log_data = {
            "request_id": request_id,
            "uid": user_id,
            "model": MODEL_NAME,
            "action": action,
            "latency_ms": int(latency * 1000),
            "tokens_input": int(tokens_input),
            "tokens_output": int(tokens_output),
            "prompt": prompt,
            "response": response,
            "timestamp": firestore.SERVER_TIMESTAMP if db else time.time()
        }
        # Log summary for quick troubleshooting
        logger.info(f"AI_FIT_MATCH_LOG [{request_id}] {action} completed in {log_data['latency_ms']}ms")
        
        # Detailed logs for investigation
        logger.info(f"[{request_id}] PROMPT: {prompt}")
        logger.info(f"[{request_id}] RESPONSE: {response}")

        # Store in Firestore
        if db:
            try:
                db.collection("ai_logs").document(request_id).set(log_data)
            except Exception as e:
                logger.error(f"Failed to store log in Firestore: {e}")

    def extract_job(self, content: str, source_type: SourceType, user_id: str = "anonymous") -> ExtractedJob:
        """Extract job details from raw content using AI."""
        request_id = str(uuid.uuid4())
        start_time = time.time()
        
        # Pre-processing based on source type
        processed_content = content
        image_data = None
        mime_type = "image/png"
        
        # Only fetch if content looks like a URL and we haven't fetched it yet
        if source_type == SourceType.link and len(content) < 2000 and (content.startswith('http://') or content.startswith('https://')):
            logger.info(f"[{request_id}] Fetching content from URL in AIService: {content[:50]}...")
            processed_content = fetch_job_page(content)
        elif source_type == SourceType.image:
            # If it's an image, we'll pass it directly to Gemini if it's base64
            if content.startswith('data:image'):
                # Extract the base64 part and mime type
                try:
                    header, image_data = content.split(',', 1)
                    mime_type = header.split(';')[0].split(':')[1]
                except (IndexError, ValueError):
                    image_data = content
            elif len(content) > 100 and ' ' not in content[:100]:
                # Heuristic: if it's long and has no spaces in the first 100 chars, it's likely base64
                image_data = content
            else:
                # Likely OCR text already extracted
                processed_content = content
                image_data = None

        # Prevent misuse: Limit content length for text
        if not image_data and len(processed_content) > 50000:
            logger.warning(f"[{request_id}] Content too long ({len(processed_content)} chars). Truncating.")
            processed_content = processed_content[:50000]

        prompt = f"""
Extraction
You are a strict data extraction engine.

Your task is to extract structured job information from raw job description text.

RULES:

* Return ONLY valid JSON
* No explanations
* No markdown
* No extra text
* Do NOT hallucinate missing fields
* If data is missing, return null or empty array

OUTPUT FORMAT:
{{
"title": "string | null",
"company": "string | null",
"location": "string | null",
"employment_type": "string | null",
"seniority": "string | null",
"summary": "string | null",
"requirements": ["string"],
"skills": ["string"]
}}

EXTRACTION GUIDELINES:

* "title" = job title
* "company" = hiring company
* "requirements" = responsibilities + requirements (merged clean list)
* "skills" = explicit tools, technologies, competencies
* Keep items SHORT and CLEAN
* Remove duplicates
* Do NOT invent skills

INPUT:
{processed_content if not image_data else "[IMAGE PROVIDED]"}
"""
        
        fallback = ExtractedJob(
            title="Unknown Job",
            source_type=source_type,
            extraction_confidence=ExtractionConfidence.low,
            summary="Failed to extract job details automatically. Please check your connection or the source content."
        )
        
        if not self.model:
            return fallback

        try:
            if image_data:
                # Multimodal request using PIL Image
                try:
                    image_bytes = base64.b64decode(image_data)
                    img = Image.open(io.BytesIO(image_bytes))
                    response = self.model.generate_content([prompt, img])
                except Exception as img_err:
                    logger.error(f"[{request_id}] Failed to process image with PIL: {img_err}")
                    # Fallback to dict format if PIL fails
                    image_part = {
                        "mime_type": mime_type,
                        "data": image_data
                    }
                    response = self.model.generate_content([prompt, image_part])
            else:
                response = self.model.generate_content(prompt)
                
            latency = time.time() - start_time
            
            usage = response.usage_metadata
            self._log_interaction(
                "extract_job", 
                prompt, 
                response.text, 
                latency, 
                request_id, 
                user_id=user_id,
                tokens_input=usage.prompt_token_count,
                tokens_output=usage.candidates_token_count
            )
            
            data = extract_json(response.text)
            data["source_type"] = source_type
            
            # Ensure required fields for Pydantic
            if "extraction_confidence" not in data:
                data["extraction_confidence"] = "medium"
                
            return self._safe_parse(data, ExtractedJob, request_id)
        except Exception as e:
            logger.error(f"[{request_id}] extract_job failure: {e}")
            # If multimodal fails, try fallback to OCR text if available
            if image_data:
                try:
                    logger.info(f"[{request_id}] Multimodal failed, falling back to OCR text.")
                    ocr_text = extract_text_from_image(content)
                    return self.extract_job(ocr_text, SourceType.text, user_id)
                except Exception as ocr_err:
                    logger.error(f"[{request_id}] OCR fallback also failed: {ocr_err}")
            return fallback

    def analyze_job(self, request: AnalyzeJobRequest) -> JobAnalysis:
        """Analyze fit between job and user profile using AI."""
        request_id = str(uuid.uuid4())
        start_time = time.time()
        
        # Prevent misuse: Basic validation
        if not request.job.title and not request.job.summary:
             raise HTTPException(status_code=400, detail="Job data is too sparse for analysis.")

        prompt = f"""
        TASK: Compare the job description and the user profile to determine fit.
        REQUEST_ID: {request_id}
        
        JOB DATA:
        {request.job.model_dump_json(indent=2)}
        
        USER PROFILE:
        {request.user_profile.model_dump_json(indent=2)}
        
        INSTRUCTIONS:
        1. Analyze the alignment between the job requirements and the user's skills/experience.
        2. Determine a 'verdict' (relevant, maybe, not_worth_it).
        3. Provide a 'apply_recommendation' (apply, apply_if_time, skip).
        4. List specific 'reasons' for the recommendation (strengths).
        5. List specific 'gaps' where the user might be lacking.
        6. Provide a concise 'fit_summary' explaining the overall match.
        7. Set a 'confidence' level for your analysis.
        
        OUTPUT FORMAT: JSON matching the JobAnalysis schema.
        """
        
        fallback = JobAnalysis(
            verdict=Verdict.maybe,
            apply_recommendation=ApplyRecommendation.apply_if_time,
            confidence=Confidence.low,
            fit_summary="AI Fit Match was unable to process this request. This might be due to insufficient profile data or a temporary service interruption.",
            reasons=["Troubleshooting: Check if your CV text is complete.", "Troubleshooting: Ensure the job description was extracted correctly."],
            gaps=["Unknown"]
        )
        
        if not self.model:
            logger.error(f"[{request_id}] Model not initialized for analyze_job")
            return fallback

        try:
            logger.info(f"[{request_id}] Starting AI call for analyze_job...")
            response = self.model.generate_content(prompt)
            latency = time.time() - start_time
            logger.info(f"[{request_id}] AI call for analyze_job finished in {int(latency*1000)}ms")
            
            usage = response.usage_metadata
            self._log_interaction(
                "analyze_job", 
                prompt, 
                response.text, 
                latency, 
                request_id, 
                user_id=request.user_id,
                tokens_input=usage.prompt_token_count,
                tokens_output=usage.candidates_token_count
            )
            
            data = extract_json(response.text)
            return self._safe_parse(data, JobAnalysis, request_id)
        except Exception as e:
            logger.error(f"[{request_id}] analyze_job CRITICAL FAILURE: {str(e)}", exc_info=True)
            return fallback

    def generate_application(self, request: GenerateApplicationRequest, user_id: str = "anonymous") -> GeneratedApplication:
        """Generate a tailored application using AI."""
        request_id = str(uuid.uuid4())
        start_time = time.time()
        prompt = f"""
        TASK: Generate a tailored job application ({request.output_mode}).
        REQUEST_ID: {request_id}
        
        JOB: {request.job.model_dump_json(indent=2)}
        ANALYSIS: {request.analysis.model_dump_json(indent=2)}
        PROFILE: {request.user_profile.model_dump_json(indent=2)}
        
        OUTPUT FORMAT: JSON matching the GeneratedApplication schema.
        
        STRICT RULES:
        1. GROUNDING: Do not make fake claims.
        2. TONE: Use a {request.tone} tone.
        3. MODE: The output_mode is {request.output_mode}.
        4. CONCISENESS: Keep it professional.
        5. SUBJECT: Generate a concise, professional email subject line relevant to the job title.
        """
        
        fallback = GeneratedApplication(
            output_mode=request.output_mode,
            subject=f"Application for {request.job.title or 'Position'}",
            email_body="Dear Hiring Manager, I am interested in this position. (AI generation failed)",
            generation_confidence=ExtractionConfidence.low
        )
        
        if not self.model:
            return fallback

        try:
            response = self.model.generate_content(prompt)
            latency = time.time() - start_time
            
            usage = response.usage_metadata
            self._log_interaction(
                "generate_application", 
                prompt, 
                response.text, 
                latency, 
                request_id, 
                user_id=user_id,
                tokens_input=usage.prompt_token_count,
                tokens_output=usage.candidates_token_count
            )
            
            data = extract_json(response.text)
            data["output_mode"] = request.output_mode
            return self._safe_parse(data, GeneratedApplication, request_id)
        except Exception as e:
            logger.error(f"generate_application failure: {e}")
            return fallback

    def generate_follow_up(self, request: GenerateApplicationRequest, user_id: str = "anonymous") -> GeneratedApplication:
        """Generate a professional follow-up email using AI."""
        request_id = str(uuid.uuid4())
        start_time = time.time()
        prompt = f"""
        TASK: Generate a professional, polite, and concise follow-up email for a job application.
        REQUEST_ID: {request_id}
        
        JOB: {request.job.model_dump_json(indent=2)}
        ORIGINAL APPLICATION: {request.analysis.model_dump_json(indent=2)}
        PROFILE: {request.user_profile.model_dump_json(indent=2)}
        
        OUTPUT FORMAT: JSON matching the GeneratedApplication schema.
        
        STRICT RULES:
        1. TONE: Professional and polite.
        2. CONCISENESS: Keep it short (2-3 paragraphs max).
        3. SUBJECT: Generate a follow-up subject line (e.g., "Follow-up: [Original Subject]").
        """
        
        fallback = GeneratedApplication(
            output_mode=request.output_mode,
            subject=f"Follow-up: Application for {request.job.title or 'Position'}",
            email_body="Dear Hiring Manager, I am following up on my application. (AI generation failed)",
            generation_confidence=ExtractionConfidence.low
        )
        
        if not self.model:
            return fallback

        try:
            response = self.model.generate_content(prompt)
            latency = time.time() - start_time
            
            usage = response.usage_metadata
            self._log_interaction(
                "generate_follow_up", 
                prompt, 
                response.text, 
                latency, 
                request_id, 
                user_id=user_id,
                tokens_input=usage.prompt_token_count,
                tokens_output=usage.candidates_token_count
            )
            
            data = extract_json(response.text)
            data["output_mode"] = request.output_mode
            return self._safe_parse(data, GeneratedApplication, request_id)
        except Exception as e:
            logger.error(f"generate_follow_up failure: {e}")
            return fallback

    def synthesize_profile(self, cv_text: str, user_id: str = "anonymous") -> Dict[str, Any]:
        """Synthesize a structured profile from raw CV text using AI."""
        request_id = str(uuid.uuid4())
        start_time = time.time()
        
        prompt = f"""
        TASK: Extract structured professional profile data from the provided CV text.
        REQUEST_ID: {request_id}
        CV TEXT:
        ---
        {cv_text}
        ---
        
        INSTRUCTIONS:
        1. Extract the user's full name, email, phone, and location.
        2. Identify their target roles based on their experience.
        3. Extract a comprehensive list of technical and soft skills.
        4. Estimate their total years of professional experience.
        5. Write a concise 2-3 sentence experience summary.
        6. Identify preferred industries based on their work history.
        
        OUTPUT FORMAT: JSON matching a partial UserProfile schema.
        Include: full_name, email, phone, location, target_roles, skills, years_of_experience, experience_summary, preferred_industries.
        """
        
        if not self.model:
            return {}

        try:
            response = self.model.generate_content(prompt)
            latency = time.time() - start_time
            
            usage = response.usage_metadata
            self._log_interaction(
                "synthesize_profile", 
                prompt, 
                response.text, 
                latency, 
                request_id, 
                user_id=user_id,
                tokens_input=usage.prompt_token_count,
                tokens_output=usage.candidates_token_count
            )
            
            return extract_json(response.text)
        except Exception as e:
            logger.error(f"[{request_id}] synthesize_profile failure: {e}")
            return {}

File: src/services/aiService.ts
import { GoogleGenAI } from "@google/genai";
import { 
  ExtractedJob, 
  JobAnalysis, 
  GeneratedApplication, 
  SourceType, 
  UserProfile, 
  OutputMode, 
  Tone,
  AIModelOutput,
  EmploymentType,
  Seniority,
  ExtractionConfidence,
  RemotePolicy,
  ApplicationMethod
} from "../types";
import { auth, saveAILog } from "../lib/firebase";

const ai = new GoogleGenAI({ apiKey: (process.env as any).GEMINI_API_KEY || "" });

const executeAI = async (
  action: string, 
  model: string, 
  contents: any, 
  config: any = {}
): Promise<any> => {
  const startTime = Date.now();
  try {
    const response = await ai.models.generateContent({
      model,
      contents,
      config
    });
    const latency = Date.now() - startTime;
    
    const uid = auth.currentUser?.uid;
    if (uid) {
      saveAILog({
        uid,
        model,
        action,
        prompt: typeof contents === 'string' ? contents : JSON.stringify(contents),
        response: response.text || "No text returned",
        latency_ms: latency,
        tokens_input: response.usageMetadata?.promptTokenCount,
        tokens_output: response.usageMetadata?.candidatesTokenCount
      });
    }
    
    return response;
  } catch (error) {
    const latency = Date.now() - startTime;
    const uid = auth.currentUser?.uid;
    if (uid) {
      saveAILog({
        uid,
        model,
        action: `${action}_ERROR`,
        prompt: typeof contents === 'string' ? contents : JSON.stringify(contents),
        response: error instanceof Error ? error.message : String(error),
        latency_ms: latency
      });
    }
    throw error;
  }
};

const cleanMarkdownOutput = (text: string): string => {
  if (!text) return "";
  let cleaned = text.trim();
  
  // Strip code block wrappers if present (e.g. ```markdown ... ```)
  const codeBlockRegex = /^```(?:markdown)?\s*\n([\s\S]*?)\n\s*```$/i;
  const match = cleaned.match(codeBlockRegex);
  if (match) {
    cleaned = match[1].trim();
  }
  
  // Check if there is introductory conversational text before the first markdown heading (#, ##, etc.) or contact block.
  // We match standard markdown headings, a bold block at the start of a line, or typical Name/Contact labels.
  const headingMatch = cleaned.match(/^(?:#+\s|\*\*[\w\s]+\*\*|Name:)/m);
  if (headingMatch && headingMatch.index !== undefined && headingMatch.index > 0) {
    const preamble = cleaned.substring(0, headingMatch.index).trim();
    // Discard conversational preambles
    if (preamble.length > 10 && (
      preamble.toLowerCase().includes("tailor") ||
      preamble.toLowerCase().includes("here is") ||
      preamble.toLowerCase().includes("cv") ||
      preamble.toLowerCase().includes("resume") ||
      preamble.toLowerCase().includes("suitability") ||
      preamble.toLowerCase().includes("focus") ||
      preamble.toLowerCase().includes("highlight") ||
      preamble.toLowerCase().includes("preamble") ||
      preamble.toLowerCase().includes("builder") ||
      preamble.split(/\s+/).length > 5
    )) {
      cleaned = cleaned.substring(headingMatch.index).trim();
    }
  }
  
  return cleaned;
};

export const aiService = {
  async extractJob(content: string, sourceType: SourceType): Promise<{ model_output: AIModelOutput, normalized_job: ExtractedJob }> {
    const prompt = `Extraction: Job Board Intelligence & OCR Engine
You are a top-tier neural data extraction engine. JSON ONLY.
OUTPUT SCHEMA:
{
  "title": "Clean Role Title",
  "company": "Company Name",
  "location": "City, State/Country or 'Remote'",
  "employment_type": "full_time | part_time | contract | internship | temporary",
  "seniority": "intern | entry | junior | mid | senior | lead | staff | principal | director | executive",
  "remote_policy": "remote | hybrid | onsite | unknown",
  "salary_info": "Currency + Range",
  "summary": "Brief summary",
  "requirements": ["List"],
  "skills": ["List"],
  "benefits": ["List"],
  "culture_info": "Vibe",
  "application_instructions": "How to apply",
  "deadline": "Deadline"
}`;

    try {
      let contents;
      if (sourceType === SourceType.image && content.startsWith('data:image')) {
        const base64Data = content.split(',')[1];
        const mimeType = content.split(';')[0].split(':')[1];
        contents = {
          parts: [{ text: prompt }, { inlineData: { data: base64Data, mimeType: mimeType } }]
        };
      } else {
        contents = [{ text: prompt }, { text: `INPUT CONTENT:\n${content.substring(0, 30000)}` }];
      }

      const response = await executeAI("extractJob", "gemini-3.5-flash", contents, {
        responseMimeType: "application/json",
        temperature: 0.1
      });

      const extractedData = JSON.parse(response.text || "{}");
      return this.processExtractedData(extractedData, content, sourceType);
    } catch (error) {
      console.error("AI Extraction Error:", error);
      throw new Error("Failed to extract job details.");
    }
  },

  processExtractedData(extractedData: any, content: string, sourceType: SourceType) {
    const model_output: AIModelOutput = {
      title: extractedData.title ?? null,
      company: extractedData.company ?? null,
      location: extractedData.location ?? null,
      employment_type: extractedData.employment_type ?? null,
      seniority: extractedData.seniority ?? null,
      remote_policy: extractedData.remote_policy ?? null,
      salary_info: extractedData.salary_info ?? null,
      summary: extractedData.summary ?? null,
      requirements: extractedData.requirements ?? [],
      skills: extractedData.skills ?? [],
      benefits: extractedData.benefits ?? [],
      culture_info: extractedData.culture_info ?? null,
      application_instructions: extractedData.application_instructions ?? null,
      deadline: extractedData.deadline ?? null
    };
    
    const normalized_job: ExtractedJob = {
      title: model_output.title || "Untitled Job",
      company: model_output.company || "Unknown Company",
      location: model_output.location || "Not Specified",
      employment_type: (model_output.employment_type as EmploymentType | null),
      seniority: (model_output.seniority as Seniority | null),
      summary: model_output.summary || `Extracted role at ${model_output.company}`,
      requirements: model_output.requirements,
      required_skills: model_output.skills,
      preferred_skills: [],
      remote_policy: (model_output.remote_policy as RemotePolicy | null),
      salary_info: model_output.salary_info || null,
      application_method: (model_output.application_instructions?.toLowerCase().includes('email') ? ApplicationMethod.email : ApplicationMethod.portal),
      application_url: model_output.application_instructions?.includes('http') ? model_output.application_instructions : null,
      deadline: model_output.deadline || null,
      missing_fields: [],
      source_type: sourceType,
      raw_content: content.substring(0, 5000),
      model_output
    };

    return { model_output, normalized_job };
  },

  async analyzeJob(job: ExtractedJob, userProfile: UserProfile): Promise<JobAnalysis> {
    const prompt = `TASK: FIT ANALYSIS. JSON ONLY. JOB: ${JSON.stringify(job)}. CANDIDATE: ${JSON.stringify(userProfile)}`;
    try {
      const response = await executeAI("analyzeJob", "gemini-3.5-flash", prompt, {
        responseMimeType: "application/json",
        temperature: 0.2
      });
      return JSON.parse(response.text || "{}");
    } catch (error) {
      throw new Error("Failed analysis.");
    }
  },

  async generateInterviewPrep(job: ExtractedJob, userProfile: UserProfile): Promise<any> {
    const prompt = `TASK: INTERVIEW PREP. JSON ONLY. JOB: ${JSON.stringify(job)}. CANDIDATE: ${JSON.stringify(userProfile)}`;
    try {
      const response = await executeAI("generateInterviewPrep", "gemini-3.5-flash", prompt, {
        responseMimeType: "application/json",
        temperature: 0.4
      });
      return JSON.parse(response.text || "{}");
    } catch (error) {
      throw new Error("Failed interview prep.");
    }
  },

  async generateApplication(job: ExtractedJob, analysis: JobAnalysis, userProfile: UserProfile, outputMode: OutputMode, tone: Tone): Promise<GeneratedApplication> {
    const prompt = `TASK: GEN APP. JSON ONLY. JOB: ${job.title}. USER: ${userProfile.full_name}`;
    try {
      const response = await executeAI("generateApplication", "gemini-3.5-flash", prompt, {
        responseMimeType: "application/json",
        temperature: 0.7
      });
      return JSON.parse(response.text || "{}");
    } catch (error) {
      throw new Error("Failed app generation.");
    }
  },

  async generateFollowUp(job: ExtractedJob, userProfile: UserProfile, originalApp: GeneratedApplication): Promise<GeneratedApplication> {
    const prompt = `TASK: FOLLOW UP. JSON ONLY.`;
    try {
      const response = await executeAI("generateFollowUp", "gemini-3.5-flash", prompt, {
        responseMimeType: "application/json",
      });
      return JSON.parse(response.text || "{}");
    } catch (error) {
      throw new Error("Failed follow-up.");
    }
  },

  async synthesizeProfile(cvText: string): Promise<Partial<UserProfile>> {
    const prompt = `TASK: SYNTHESIZE PROFILE. JSON ONLY. CV: ${cvText.substring(0, 10000)}`;
    try {
      const response = await executeAI("synthesizeProfile", "gemini-3.5-flash", prompt, {
        responseMimeType: "application/json",
      });
      return JSON.parse(response.text || "{}");
    } catch (error) {
      throw new Error("Failed synthesis.");
    }
  },

  async generateTailoredCV(job: ExtractedJob, userProfile: UserProfile): Promise<{ markdown_content: string }> {
    const prompt = `
CANDIDATE SUITABILITY & KEYWORDS:
Job Title: ${job.title}
Company: ${job.company}
Required Skills: ${job.required_skills?.join(", ") || "None specified"}
Job Summary: ${job.summary || "None specified"}
Job Description Text: ${job.raw_content || ""}

MASTER CV DATA:
${userProfile.cv_text}

MISSION:
Carefully analyze the job details above and customize the master CV to align perfectly with this role. Adjust accomplishments, highlight the most relevant skills, and optimize the ordering and phrasing of bullet points to match the job's keyword density while preserving absolute truthfulness.

STRICT OUTPUT CONSTRAINTS:
1. Output ONLY the raw Markdown of the final tailored resume.
2. DO NOT include any conversational preamble, introduction, explanation, greeting, or conversational chatter (e.g. do NOT say "To tailor your CV for...", "Here is...", etc.).
3. DO NOT include any note or explanation at the end.
4. Begin directly with the resume header (typically "# Name" or "Name").
5. The resume MUST be professional, ATS-optimized, and use clean markdown syntax.
`;
    try {
      const response = await executeAI("generateTailoredCV", "gemini-3.5-flash", prompt, {
        systemInstruction: "You are an elite, professional resume-writing system. You only output raw, ready-to-use resume markdown. You are strictly forbidden from writing any introductory greetings, transition sentences, commentary, explanations, pre-text, or post-text. You must output the resume directly.",
        temperature: 0.3
      });
      const cleaned = cleanMarkdownOutput(response.text || "");
      return { markdown_content: cleaned };
    } catch (error) {
      throw new Error("Failed tailored CV.");
    }
  },

  async generateMasterCV(userProfile: UserProfile): Promise<{ markdown_content: string }> {
    const prompt = `
CANDIDATE INFORMATION:
Name: ${userProfile.full_name}
Title: ${userProfile.target_roles?.join(", ") || "Software Engineer"}
Location: ${userProfile.location || "Nairobi, Kenya"}
Skills: ${userProfile.skills?.join(", ") || ""}
Experience Summary / Raw Text:
${userProfile.experience_summary || ""}
Original CV:
${userProfile.cv_text || ""}

MISSION:
Synthesize a highly polished, comprehensive, and chronologically organized Master CV in beautiful markdown format using all available user profile and experience data. Ensure achievements are quantified (using impact/metrics where possible).

STRICT OUTPUT CONSTRAINTS:
1. Output ONLY the raw Markdown of the final Master CV.
2. DO NOT include any conversational preamble, introduction, explanation, greeting, or conversational chatter (e.g. do NOT say "To create your Master CV...", "Here is...", etc.).
3. DO NOT include any note or explanation at the end.
4. Begin directly with the resume header.
`;
    try {
      const response = await executeAI("generateMasterCV", "gemini-3.5-flash", prompt, {
        systemInstruction: "You are an elite, professional resume-writing system. You only output raw, ready-to-use resume markdown. You are strictly forbidden from writing any introductory greetings, transition sentences, commentary, explanations, pre-text, or post-text. You must output the resume directly.",
        temperature: 0.3
      });
      const cleaned = cleanMarkdownOutput(response.text || "");
      return { markdown_content: cleaned };
    } catch (error) {
      throw new Error("Failed master CV.");
    }
  },

  async generateCoverLetter(job: ExtractedJob, userProfile: UserProfile): Promise<{ markdown_content: string }> {
    const prompt = `
TASK: ELITE GENAI ENVOY ARCHITECT
Style: Technical, Direct, High-Impact, Builder-Centric.
Role: A world-class engineer drafting a technical partnership proposal.

CANDIDATE: ${userProfile.full_name}
LOCATION: ${userProfile.location || "Nairobi, Kenya"}
TECH STACK: ${userProfile.skills?.join(', ') || "Full-stack Engineering"}
FULL EXPERIENCE: ${userProfile.cv_text?.substring(0, 3000)}

TARGET ENTITY: ${job.company}
ROLE: ${job.title}
JOB INTELLIGENCE: ${job.raw_excerpt?.substring(0, 1500)}

MISSION:
Do not write a generic application. Propose a technical solution to the specific challenges implied in the Job Intelligence. 

REQUIRED ARCHITECTURE:

1. HEADER: Name, Location, Contact.
2. DATE: [Current Date]
3. RECIPIENT: Hiring Manager at ${job.company}.
4. SUBJECT: **RE: ${job.title}**
5. THE HOOK: A high-signal opening that connects your background in GenAI/Backend directly to ${job.company}'s mission. Avoid "I'm applying for...". Start with: "I build production-ready ${job.title}-related systems..."
6. DYNAMIC PILLARS (3 Sections): Identify the 3 most critical technical needs in the job description (e.g., Automation, Scalability, Security, UI/UX, multi-tenancy). Create 3 bold headers and describe exactly how your specific history makes you the solution for those 3 needs. Use data/metrics from your CV.
7. CULTURAL & LOCAL SYNC: A short paragraph on why ${job.company}'s specific engineering culture fits your "builder" mentality. Mention being Nairobi-based for fast iteration if applicable.
8. THE PROACTIVE CLOSE: A confident request to discuss their current technical roadmap and how you can add immediate value.

STRICT CONSTRAINTS:
- ABSOLUTELY NO: "enterprising", "landscapes", "tapestry", "digital age".
- Use bold only for Subject and Section Headers.
- Word count: 280-350 words.
- Tone: Senior Engineer to Senior Engineer.
- Format: Markdown only.
`;
    try {
      const response = await executeAI("generateCoverLetter", "gemini-3.5-flash", prompt, { temperature: 0.6 });
      return { markdown_content: response.text || "Letter failed." };
    } catch (error) {
      throw new Error("Failed cover letter.");
    }
  },

  async verifyCVStructure(markdown: string): Promise<any> {
    const prompt = `TASK: VERIFY CV JSON. CV: ${markdown}`;
    try {
      const response = await executeAI("verifyCVStructure", "gemini-3.5-flash", prompt, { responseMimeType: "application/json" });
      return JSON.parse(response.text || "{}");
    } catch (error) {
      return { success: false, score: 0, logs: ["Offline"], pillarStatus: {} };
    }
  }
};
